### ISSUE TITLE
Proposal for an event system

### ISSUE THREAD

**Issue by vlegoff** [evennia/evennia#1188]
#### Open for discussion

If this feature is judged interesting or valuable, I offer to code it and put in contrib, so individual users could decide to use it or not.  So this issue is more a "proposal to be reviewed" than a request.

#### Brief summary of feature

I would offer an event system that would:

1. Allow to customize individual objects' BEHAVIOR.
2. Not ask to edit a single line of code in the source code of the engine.

I would offer to add:

1. The EventHandler (see below for details).
2. The @event command to edit/remove/list events on a specific object.

#### Rationale

Offering such a feature would allow builders (or immortals) to add features to the universe without bothering about the code at all.  I know a lot of builders who have basic understanding of Python, but are not confident enough to tickle Evennia or the full game engine to add features.  Furthermore, there are cases when one only wants to add a single feature to a single object without modifying the behavior of others.  Doing so without creating a custom class and overriding some hooks seems difficult, and not really friendly to these builders.

Alternatively, a lot of engines offer ways to "script" individual behavior, but these "scripts" are pretty limited.  I propose to keep these scripts in pure Python and allow about everything Evennia would allow normally.  Using the @event command (see below for details) would allow builders with limited skills in Python to add features to their creation, without limiting others, more skilled, to add complex features.

Good use cases COULD be:

1. I want an exit to prevent characters from passing through if they haven't finished a quest.
2. I want a NPC to act specifically when a character arrives in his/her room (including say, emote and so on).
3. I want a NPC to move at regular hours in and out of his/her shop, to follow a path made of several rooms and walk at a specific speed.
4. I want a room to prevent characters to say something, but allow them to say something else and possibly react to a keyword.
5. I want to create a single object that damages all characters in the room if dropped (just an object, not a type, not a typeclass).
6. I want a room to update its description at various moments of the day, the weather, the season and other random factors (like the character looking at it having completed a quest).

There are a lot of other possibilities.  All that could be done with some editing of the code (either complex or simple).  As far as I know, none of that can be done, for the time being, without editing the source code.

##### A single contrib

My idea is to put this feature in a contrib.  Game developers who want to use this system would just have a few modifications to do in their code (adding the command and replacing the parent DefaultObject should be enough).  Where to store these events is still under discussion: it could be either in the object database itself, or in a remote table.  Both have advantages and drawbacks.

##### A command to manage events

For builders or immortals, the idea would be to use a single command, @event.  This command would look like this:

    @event [object] <event name>

With switches:

- /edit: open a line editor to edit the script, with additional features to handle indentation.
- /remove: remove this event and the code it is associated to.

The object would either be found in the room itself or accessed through a #id.

A quick example of manipulation:

    @event/edit north traverse
    (Open the line editor to edit the event's code for the exit north in this room)
    if character.db.health < 100:
        character.msg("You're not feeling well enough to go there.")
        deny()
    elif character.db.health < 30:
        character.msg("You're feeling much too weak to go that way.")
        deny()
    (Leave the editor with a custom command)

And then when you take the 'north' exit in this room, you will be blocked (with different messages) if your health isn't above 100.

Note that I put the indentation here for landmarks.  In this editor, moving through indentation would be automatic (but could be changed through alternative commands).

deny() is a function that would be created just for these events.  For the time being, it raises a custom exception, so interrupt the script altogether, and signal the caller (method Exit.at_traverse in this case) that the operation is to be cancelled.  Not all events could deny() (some may be called after a certain operation has been processed).

##### An EventHandler

A possible design would be to create a class with multiple inheritance: ObjectDB (or even DefaultObject) as a first parent, and EventHandler as a second, to include the few methods added by the ScriptHandler (like call_event).

All objects inheriting from this class would be able to access both the method of objects and the additional methods to call events.  There are other ways to handle the situation, of course.

##### Reviewing

So what do you think of this suggestion?  Please comment as will, I'm not coding anything before obtaining positive feedback.  Although, if this idea sounds dumb to every other developer, I'll create it for my own MUD!

Thanks for reading,

**Comment by volundmush** [comment]
This is a really cool idea. I have no idea HOW it would be implemented, but it reminds me of the 'mobprogs' that I used to write when I helped out at a Dragonball MUD based on SMAUG...

**Comment by BlauFeuer** [comment]
[Niflheim's effects ](handlerhttps://bitbucket.org/whizzlenizzle/niflheim-lib/src/201c086c054f382bf7f20f47d411e808747173a9/niflheim/core/characters/effects.py?at=master&fileviewer=file-view-default) from [The Niflheim Project](https://niflheimproject.wordpress.com/) uses scripts which do allow for arbitrary code to be run, but supplies no editor for it or any permission system for what can apply effects to other objects, but it already has a good foundation structure for an effects handler and would be nice if someone picked this up and continued it.

Adding your `@effects` (similar to the `@ticker` for tickers) command would be nice to review and edit script triggers and their effects.

**Comment by Griatch** [comment]
First of all, thanks for a comprehensive and well-written feature request/proposal! I think it may be worth to attract the attention of @Kelketek  at this point since he has been working on his *collab* contrib system for many years and it also has some sort of softcode in it lately; so hopefully he can chip in on the time line and how much something like this would overlap. 

So onto this proposal. There are some issues I must point out: 

The most prominent problem with this is security. What you are suggesting is essentially offering builders the ability to run arbitrary Python code. This is a *severe* security problem that may lead to a potentially malign builder compromising your computer/server. As much research has shown, Python's awesome introspection is its enemy here: it is essentially impossible to secure Python against gaining access to things like file I/O and to create infinite loops that kills the server (yes I have spent a lot of time with this, this was the ill-fated Evlang expansion). The infinite loop could maybe be resolved by running builder scripts in a separate killable process, this however suggests a process pool and introduces the problem of race conditions if you try to access the database.

I don't want to discourage you, but security is a big deal and I think the suggested implementation would at least need to acknowledge this. I guess it might be okay for a small and tight team of friends co-developing (everyone knowing and trusting each other) but in that case they can just as well do it in Python directly. In practice many game builders are strangers from the internet or adopted from the player pool and should not have that kind of power over your server. For this to be safe the system facing the builder needs to be restricted. 

This is why the lock- and inlinefunc systems are designed the way they are. With the example of the  lock string, it is itself modifiable by a builder and only supports very basic and safe AND/OR/NOT logic. The complexity is in the lock functions which a trusted developer creates in Python for builders to use. I can't help but think something similar would need to be done for your `@events` system for it to be safe. 

Here are some comments on the use cases you mention. 

 1. Locking an exit. This is possible by for example `@lock exit : traverse:attr(latest_solved_quest, rat_attack)`. A developer could make a lock function specifically for the use of handling a quest (`@lock exit : solved_quest(rat_attack)`). This could of course be gathered under a single command like `@event` if one prefered. 
 2. NPC reacting when entering a room: This is today done by the developer by overloading hooks on the room. This could in principle be offered as a custom room managed through a lock. 
3. NPC moving along a given path: This could be done by the builder offering a series of room names in a sequence, like `@events move npcname = north, east, south, west, north, ...`
4. A room preventing a player to say something: This would be a custom room implementing the `at_say` hook. This hook could mute the speaker if, say an attribute or a lock is set by the builder. 
5. Objects that do custom things is a tricky thing, full control of this suggests softcode, which is a whole different can of worms. The recommended way to offer features like this would be for the developer to offer a typeclass with "slotted" functionality that the builder can activate. For example, it may have a value `damage_room_when_dropped` that defaults to 0 but which the builder can set. It could by all means still be manipulated with a joint `@event` command. 
6. Room updating is already available in e.g. the Extended Room typeclass. More features could be offered in a slotted way as suggested above. 

The traverse example you give, locking the exit based on a series of conditions, is of course easy to do in a typeclass; from in-game it might *possibly* be done using custom lock functions but those are not intended for advanced logics and it would quickly get very clunky. In reality it is something that is hard to imagine doing without offering Python access (with its security issues) or some sort of custom softcode (which tend to be clunky and means another language to learn). Evennia thus recommends that such more advanced features be implemented by the developer in the typeclass and make the system controlled by flags safely set by the builder. 

That said, I could potentially see this as a nicer presentation around the `@py` command intended for `Immortal`-level builders.

**Comment by volundmush** [comment]
As much as I curse softcode for PennMUSH and RhostMUSH, it does have its merits. It shines when used for VERY SIMPLE THINGS along the line of mobprogs, locking commands, making an NPC speak, building atmosphere, and that sort of thing. That is what most of the stuff on MUSHcode.com aims to accomplish. What I do with it is thoroughly insane.

If Python is too insecure to grant builders, then... let's not do that. :D But (as Kelketek's project shows) it shouldn't be too hard to come up with a simple interpreter that has its own rules about permissions and security. ... well I say that, but it's still a lot of work to actually make it happen.

**Comment by Kelketek** [comment]
The code in progress is on the collab branch. I think it may need a rebase, but that might be a good starting point to look at. If it looks likely to already cover the things you need, I'd say either wait for that or help me build it-- my time is much more limited these days, but most of the hard work is already done for collab. It needs some glue pieces now surrounding success/failure messages, script handling (I'm not sure-- I might have fixed that already) a few more commands and tests, and then review/addressing notes.

If you're interested, I'm happy to spend some time chatting with you to help you find ways to contribute to it.

**Comment by vlegoff** [comment]
Thanks, first and foremost, for your comments.  I must say I thought the proposal would be highly disliked for its added layer.  I'll try to address points and concerns here.

The most important issue, and I should have addressed it in the proposal itself, is security.  There's no denying it: having arbitrary Python code in action can be dangerous.  Can break everything, and I'm not only talking about the game.  In theory, the persons who have been trusted to execute @py could easily be granted such responsibility.  In practice, it might be a very small number, and would indeed diminish the usefulness of the feature (as some pointed out, if restricting that much, why not code? it's likely users with the @py function will know how).

There's another possibility which is to restrict what the user can actually use.  There are some implementations of "Python sandbox" around here.  There is also the possibility to run every single added line of Python through an analyzer that would just prevent some syntax, some function calls (like everything outside Evennia's library).  We could go as far as providing "safe" objects to represent the universe and only have builders interact with a small number of features, but that might limit the usability as well.

I was also thinking of a post-treatment measure: no restriction on what can be entered, but every line is logged and recorded outside of the MUD, and the one controlling the server can review these lines, to check that his/her builders didn't do something like building a back door or sending data to other sources or use sys.exit... it would allow to control when the server crashed, and remove builder right if it should happen.  This is not a perfect solution, of course.

What I have on my current MUD is a custom-made scripting language that converts each instruction in pure Python, but with some rules.  This parser is a bit limited (though I haven't seen as many simple and powerful scripting languages in MUDs out there).  Admittedly, it would require a lot of code to ensure the user does what has been planned and no more than that.  That's an option.

@Kelketek: I'm still interested by alternative codes and solutions that provide this feature.  I truly believe this event system (or whatever it's called in the end) would be great for quick customization without code editing.

**Comment by Kelketek** [comment]
@vlegoff I think you could implement an event system using the collab branch. Since the collab branch has the ability to store little 'softcode' snippets and run them, all you'd have to do is have a place for the user to store these scriplets, and a way to track when an event happens and what things to call when it does, which could be done using some global Singleton Script (as in the database object) that does message routing and checks to see what scriptlets should be called using what parameters when an event happens.

**Comment by Griatch** [comment]
> There's another possibility which is to restrict what the user can actually use. There are some implementations of "Python sandbox" around here. There is also the possibility to run every single added line of Python through an analyzer that would just prevent some syntax, some function calls (like everything outside Evennia's library). We could go as far as providing "safe" objects to represent the universe and only have builders interact with a small number of features, but that might limit the usability as well.

I have optimistically gone this route in the past - whitelisting and blacklisting to try to make a "safe" subset of Python to use. If you don't trust me, trust the hundreds of others on the internet saying the same thing - it's not a route you should try. Even with the most restricted subset there are ways to create an infinite loop that eats up all your memory or access the file system. Running things in a root jail with low priority will help save the rest of your machine but  still leaves your server potentially vulnerable. It was/is certainly not something I was willing to put up as an official Evennia code for people to burn themselves on. 

> I was also thinking of a post-treatment measure: no restriction on what can be entered, but every line is logged and recorded outside of the MUD, and the one controlling the server can review these lines, to check that his/her builders didn't do something like building a back door or sending data to other sources or use sys.exit... it would allow to control when the server crashed, and remove builder right if it should happen. This is not a perfect solution, of course.

As long as the code is not executed until after it has been vetted, this is probably reasonably safe. Of course, this way the builder has no immediate feedback on their code (can't easily fix bugs) and they could just as well commit a PR against your development repo since you have to vet it anyway. 

> What I have on my current MUD is a custom-made scripting language that converts each instruction in pure Python, but with some rules. This parser is a bit limited (though I haven't seen as many simple and powerful scripting languages in MUDs out there). Admittedly, it would require a lot of code to ensure the user does what has been planned and no more than that. That's an option.

This is what is called a softcode language. It is probably okay out of these options. The problem lies with the fact that it will never be Python, but will open up a whole new syntax to support, security holes of its own to manage and performance to optimize. It's also tempting to start using it for more than it was intended for. I still think that offering your builders powerful build tools rather than a full script language is the way to go, personally. But I'm open to discuss, by all means.

**Comment by vlegoff** [comment]
@Griatch: I have uncovered the same truth from recent research (and I didn't have to go very far to see the "WARNING, unstable, unreliable, unsecure" sandbox.  So this solution is obviously out.

I tend to trust my builders... but to what degree?  That's obviously for anyone to decide.  My initial thought was to have these events validated by an admin at first, then, if the builder seems reliable, authorize him/her greater access (the possibility to add events without validation), BUT an admin will always be watching to see what has been done.  That's the point though: **what has been done**.  A builder that has gone through the initial validation and proven reliable for some time (depending on the MUD policy) will still be able to make a lot of damage once he/she's given additional power.  However, if admins keep a regular watch over the events that have been provided, and remain in control of events (running the server while disabling all events, should one create a crash, for instance), most problems might be avoided.  Spotting a back door or some malevolent code is, after all, not that complicated in such a language.  It's like granting commit access but forbidding users to commit to a certain branch: I have never done so, and some developers (like myself) believe that when users are trusted to do whatever please them, they usually play by the rules much more willingly.  That's a personal consideration, obviously, that shouldn't interfere with anyone's development.  That's why I would keep it as an optional tool, in any case.

To conclude this comment, I don't like 'softcode' that much either.  It opens the door to lots of additional problems and doesn't allow as much flexibility.  Some will (and have) argued that it all depends on the quality of the softcode.  Perfectly true.  But after my experiences of creating such a system for eight years, I believe one always ends up with trouble and limits, no matter how often the softcode is updated.  Again, that's a personal consideration.

**Comment by Griatch** [comment]
@vlegoff 
In principle one could at least allow Immortals to add Python code snippets to an object and execute that on demand with no concept of security (that would be fine for Immortals, they have access to the `@py` command anyway). But again, they should have access to the proper external code tools anyway.

**Comment by vlegoff** [comment]
That's why I would be in favor of configurable access rights even to builders, customizing the objects/rooms/characters they create, for it's almost the only interest, having powerful creating without asking to change the source code.  Now, if you like it, you activate it and you configure the rights (some can't, some can with validation, some can with post-validation, some can without restriction).  I don't know how that can be set, but I'm sure that can be, with the current system.

**Comment by Griatch** [comment]
You could change your game to allow more Immortal-level commands to Builders if you wanted. Or you could just give builders Immortal-level access off the bat if you are feeling adventurous. As said, the main way to make an object configurable by a builder is to create your typeclass such that features of it is settable by the builder (a slot-like system). Beyond that you need to either go the softcode- or the python-scripting route with all that entails.

**Comment by BlauFeuer** [comment]
"Proposal for an event system" would be more approachable if it proposed just an event system - comprised of an event handler that triggered an event when an object met certain criteria - beyond what can already be done now with locks, lock functions, and inline functions.

Even if the events and criteria initially required coding extensible modules, at least builders could choose from a hopefully growing list of triggers and events.  As more use cases are discovered and more creative feedback from builders is given, the selection and possibilities would quickly grow.   Commands to choose triggers and events and plug in a builders choices, along with some descriptive documentation would make it a more approachable system for all, and the more general it is, the more games would want it.  A library of events and triggers would be shared, too - perhaps in a similar way that people have pride in their softcode creations and share them.

**Comment by vlegoff** [comment]
Okay, I'll code something and perhaps create a PR just for review... or perhaps there's a better way of doing it.
@BlauFeuer: if I understood correctly, that's what I had in mind all along.  Some triggers (events) would be linked to already-existing objects.  But, if a developer adds the eat command, it will be easy to create the "eat" or even "can_eat" events and connect them.  Perhaps, if enough people are interested, some events would be shared as you said, it will depend on whether I can design something simple yet powerful.

**Comment by BlauFeuer** [comment]
I think it's interesting and would be valuable; I just sense that how its implemented will determine how extensible it is - if it's simple yet powerful.

Also, I'm hopeful that it will be a system that uses some of the API that we already have, like monitor handler, locks ("can_eat" sounds like a lock), lock functions,  script typeclasses that _can_ run arbitrary code and provide a trigger that reoccurs n-times after the initial event, and generally gives builders ability to edit their own creations to give more depth to the setting so it's more dynamic.

I look forward to seeing how you approach prototyping a test case and if it does grow out of a slot-like system or jump directly into softcode scripts or writing arbitrary Python code directly.

**Comment by vlegoff** [comment]
@BlauFeuer: once more, I may be mistaken, but if I turn the "can_eat" into a lock, then the message will always be the same and the reasons won't vary... except, that is, if you edit the code.  I was thinking of having them as triggers (events), the can_* to check that the command can be used (it won't follow the syntax of commands, there will be other actions, like can_traverse) but in most cases (like can_say).  Obviously, I need to have a system that will remain open to additions and will probably be very sloppy at first, since I may not be really familiar with some of Evennia's features, not in depth, anyway.  So the first proposal (probably the first proposals) will just be open for review and discussion, not for immediate integration.  This is not a small feature and it could so easily lead to latter disaster.

**Comment by BlauFeuer** [comment]
Huh? The lock doesn't store the message; it just tests something and returns a fail/success.  Messages aren't inherently in locks, and not normally in lock functions unless you put them in.

You may be mistaken.  To mitigate the chances of later disaster, check out some of the existing features in Evennia so you're not having to re-invent parts of it that have proven useful so far.

**Comment by Griatch** [comment]
As Blau points out, a lock is not a fixed condition which is the same every time. Depending on what the  `can_traverse` lock function looks like, it's result can be very different depending on who use it. Here is an example of a lock limiting traversal of, say, a door:

```python
"traverse: holds(green_key) or tag(locksmith) or perm(Builders)"
```

Anyone holding the green key, is tagged as a locksmith or is a builder will get info the door. Builders can safely create locI 's like this in game and you can supply more complex lock functions for them to use if you like.

**Comment by vlegoff** [comment]
Hi everyone,

I've finally finished the documentation (this is obviously a first draft, to see what could be done with this system).  Nothing is developed yet, this is just a new stage of the proposal.  It helps me clarify how to handle the issue.  If you have time to review it and give me your comment, I would be ever so grateful.  https://github.com/vlegoff/evennia/tree/events/evennia/contrib/events

Enjoy!  And thanks in advance,

**Comment by Griatch** [comment]
@vlegoff 

That's a comprehensive text, thanks for elaborating on your thoughts. Here are my comments. 

- First of all, let me say I like the concept as described here. Giving immortals more in-game coding capabilities without needing softcode is appealing. 
- You need some different word to separate the event from the code to run on the event. I'll call it "the event" and the "event callback" for now. 
- I know this readme is mainly to help gather your own thoughts, but be aware that this is way, way too long for an eventual future contrib readme. Your audience for a contrib is the developer, not the in-game player. If you want to write the user docs, you should mark that in a separate section and instruct the dev to paste that into a help entry. It's also not the contrib's job to repeat how standard Evennia features like `#dbrefs` and permissions work, focus only on the things unique to the contribution. 
- The installation would be more flexible if you are simply told how to install the things you need - like "import and add the CmdEvents command to the Character Cmdset", inherit these typeclasses, these are the new options you could add to your settings file, etc. This makes it easy for a knowledgeable dev to mix and match multiple contribs. This also puts it in line with how other contribs do this. 
- The "Event Editor" would be useful as a separate PR - a new "programming mode" of the EvEditor that is set programmatically when launching the editor to operate on Python code. This would then have stand-alone use for adding an `/edit` switch to the `@py` command.
- I really like the permission levels with validation or not, this seems to solve some of the problems for security-concious people. If the Python is not to be run, I think the programming-EvEditor mode should include a linter like pyflakes though, so common indent or syntax errors could be debugged directly so as to not offload on the immortal to debug simple things as well as validate. I wonder if the builder should be able to write some simple test case too, like if input -> expect this output? 
- Ditch Memories. It just introduces yet another mini-system to learn only for events. Builders have access to Attributes already today, that's one of the things they are designed for. The state of Attributes can also be monitored with the MonitorHandler. 
- Speaking of the memories, I don't understand your tea example. You do `@event cup of tea = create` and then make code for setting things. But `@event` operates on objects, and `cup of tea` is not yet created you say, so how could you assign an event to it that fires when it is created? If `cup of tea` is a typeclass, how would you store that setting? Either way, it would be better to use what is already there, namely Attributes. 
- The tea example seems a bit unintuitive over all. `del_memory` is first described as the name of an event, but in the code `del_memory(object, "temperature", 20)` seems to be a timed delete function, deleting in 20 seconds, while `add_memory(object, "temperature", value) adds `value` immediately`? 
- I like the "do this at 12:00" concept. Such cron-like functionality should probably be added to Evennia's `gametime` module as a separate PR, it would be of general use outside the Event system. 
- I don't understand the motivation for giving events `chain_` names. If you are using `call` to execute other events, why not just call them by the name of that event? It's like having a Command calling another Command. The user supposedly have access to all events anyway?
- Overall, I miss information about just what is available in the local namespace of the event callback.  Is `evennia` available? Are only the helpers available? How do you access the object the event callback is defined on? Is the original creator of the callback available? Example code for how the event call happens would be good. 
- Many of the "helper" functions seems to be just thin wrappers over existing functionality. Be careful here to not overdo this and end up with a whole separate API that you must maintain and update. We have a lot of documentation and help for the resources Evennia makes available, you should avoid to create a parallel syntax where you can avoid it - if you do you are veering into softcode in the sense that you are creating code useless outside the event system. If people are coding in Python anyway, they can just as well learn the real functions from the flat API. If something is too hard to use, that's a usability issue against Evennia. 
- Events in descriptions are already handled by`$inlinefuncs()` which are considerably more powerful and allows nesting too. Inlinefuncs are nice in the way that they are only evaluated when someone actually reads the text in which they are - and can be different depending on who is looking. In this example you'd just write your own inlinefuncs to check the event state when a character views the description. Inlinefuncs [are described here](https://github.com/evennia/evennia/wiki/TextTags#inline-functions).
- Adding of new events through a module customized in settings look good and Evennia-esque. 
- Disabling all events could more clearly be done with a setting, `EVENTS_ENABLED=False`. That could retain the `@events` command but make sure to disable the event execution. 
- I am missing implementation details of just how things are stored and how things are executed under the hood - especially how events are tied to the Evennia engine, which is the important bit at this stage. 

As far as I can tell, even though it's explained with a lot of detail this is actually a pretty simple concept and I'd like if it could be kept that simple. Lemme see if I can summarize this the way I understand it: 

*The Event system contrib allows privileged users (Immortals by default) to write their own Python code snippets in an in-game code editor. This code is stored as text on an object and keyed to a fixed set of game events defined by the developer. When the given event occurs, the code is loaded and executed in a name space including common utility functions. Security is based around trust and potentially manual vetting of code before execution.*

Does that sound about right?

**Comment by vlegoff** [comment]
@Griatch: first, thanks for reviewing the help file.  I know it's not really short.  Here are my answers to your comments.

- A word to distinguish between events and event code: events and event callbacks are fine with me.  Perhaps we should keep events for the code, and call the system triggers... I don't know if it makes more sense.
- The README is definitely too long.  But it seemed more sensible to write it in a separate file rather than the issue.  In theory, this README could (after some changes) become a user documentation on the wiki... although some parts are too technical, whether others aren't described enough.
- I'll be clearing the instructions to install, with everything.  In my basic plan for development, the `@event` command was automatically to the character CmdSet, but that might be counter-intuitive (sometimes, additional lines of codes are better for clarity).  Changing inheritance of typeclasses might not be necessary either, but once more, that's a matter we'll have to discuss, it might lead to misunderstandings.
- Would it make sense to have, say, a keyword like `code=False` to the EvEditor?  I was thinking of creating a new class inheriting from it, but I see the point of grouping them for the `@py` command for instance.  It would add a few lines to the editor, but perhaps not that many.
- Good that the permission system sounds elaborate enough.  We'll have to discuss (probably on chat) ways to simplify code of users who's work isn't automatically accepted.  Since there's a chance that these users be beginners in Python, more advanced debugging options make sense, although I'm not sure how to offer them.
- Easy to remove memories.  I'm a little hesitant, because plying directly with attributes could raise errors that would be hard to debug, in case of conflicts.  There's also the fact that memories would be monitored and this monitor would be linked to events, that makes some things easier to develop.  But it can very well be put in a separate contrib... or even only for my own use, if you really don't approve :) .
- The tea example was a bit misleading.  The `create` event, as you pointed out, had been called when the object was first created or spawned.  This is something I haven't really figured out with Evennia.  I took this example because it was readily-available, but I shouldn't have in this case.
- I was thinking about that regarding the in-game time yesterday.  It seems the functions in utils contain the basic analogies... though I don't know if you can create alternative calendars, like a year with 10 months of 30 days in them.  In any case, I thought of adding a function in this file, say "program" or "schedule" that would create, behind the scene, a script that would call a callback at a defined, game time.  This function would take as argument the time, the callback and alternative arguments to pass to the callback.  Since it's a script, it will be saved normally (given the callback is a top-level function).  However, when I thought of it, I found it limiting that I should only think about 24-hour usages.  You cited the cron program, I was wondering if we shouldn't offer something more precise to do certain actions like twice in a game day, or every week, or every month, or whatever.  This might be a bit more complex and I don't really know if the system of game time is meant to go that far.
- Chain names aren't really useful.  By default they would just be numbered, chain_1, chain_2, chain_3 and so on.  I just thought that, sometimes, a developer could wish to give a more explicit name.  That's far from mandatory and not really useful, except if the developer comes with more than 3 chained events in the same object... which might be rare, I don't really know.  chain_X events have this name, because they are not triggered by the system.  Sure, the `call()` function could be used to call other events, but then, the system could do, which might lead to conflicts (the example I gave had an event to do a simple thing as a result of another event).
- I didn't think much of namespace, you're right.  A good idea might be to have all the helpers defined in the namespace, along with some of Evennia's readily-accessible features... perhaps (only perhaps) the typeclasses... and some utils functions like create_* and search_*.  I don't know if that makes sense to include more, after all, users will still be able to import something if they want, say, to display an EvTable in their event for bowling contests :P .
- Regarding helper functions, I admit I took a shortcut because I don't often use the create_* functions.  But we could definitely provide them, as you pointed out, it will be more logical and we won't create an API over what already exists (and what already exists isn't limiting :D ).  If the developer adds a typeclass, he/she might have to create the corresponding helper functions to search or create.  Also, my example `station = room(id=22)` wasn't great, but I wonder how to do this simply... the syntax `station = Room.objects.get(id=22)` might be a bit too much for users of the system.  Or perhaps not.
- I examined inline functions before talking about events in description.  The reason why I have included them is that I don't think inline functions are easy to read when it comes to multiple variables.  For instance, the example I gave almost at the top of the file, of having a slight portion of the description changing if it's between 4-7 AM, or -10 AM, or 10-12 AM, or 12-15 PM... that sounds like a heavy syntax for inline functions.  Perhaps it's something I haven't been studying well enough, however.
- I quite agree for disabling all events: a setting should be more than enough.  It's not like the function could take arguments anyway, and it will avoid modifying the startup hook.
- One of the reasons why I haven't detailed how things are stored and defined is because I'm still hesitant.  I wanted to have a user documentation that would remain mostly the same no matter the path I selected.  One way of storing events would be to have an attribute `_events` (or even èvents`) in objects, with as key the event and as argument, the list of events.  Most of the implementation will be static anyway, and described per class (the list of events this typeclass implements and use, for instance).  It could be loosely stored in a class variable, since it won't need to be saved by the database.  Or they could be some lose matching between typeclass and event list.  This still remains to be thought through.

Regarding your summary, it seems to be quite accurate (and much shorter than what I had come up with).  I don't think the implementation would be too hard to code either, I just mean to do it right and to have a plan that will still holds true in latter versions.

Thanks again for feedback,

**Comment by Griatch** [comment]
Some replies: 

- As for the code editor, yes something like a keyword `code=False` was what I had in mind. 
- There are python linters like pyflake that could run on code and warn about errors. This would be an optional dependency; the person using the event system could be asked to install it with `pip` (or maybe not having it will just disable the checking). Again, this could be useful for the expanded code-EvEditor as well.
- You can customize the calendar produced by the gametime module by setting the time factors and steps for the various calendar units. 
- I feel the `chain_` names are just adding complexity here. There is nothing stopping you from having an event that Evennia never actually checks. I think what you are looking for here is the ability for the immortal to create custom callbacks to be used by other parts of their code; not sure this is a good idea: It adds complexity and makes the code harder to vet and follow (remember that this is code stored in strings on objects, not easily contained in modules). This encourages people to build large systems of interchanged code in-game which seems to be counter to the intent of this system. 
- One can certainly discuss which helpers should be available. A search_ method from Evennia's API does not have to be unique to the typeclass: `search_object()` takes the typeclass as first argument. 
- You can use inline functions in the way that you hide away all the complexity inside the function (which is under your control). The only difference to your example is that it would look like `$function()`, rather than `$function` as you showed. It sounds like a good idea not to have two different syntaxes to remember.

**Comment by vlegoff** [comment]
Replies to replies (the list is getting shorter :D )
- Okay, I'll work on a PR for the code EvEditor.
- I'll take a look at `pyflake` or other alternatives, I didn't even know it existed.
- Regarding the calendar, I spoke too quickly and was able to create a custom calendar yesterday.  After all, with the help of just seconds, one can build something quite complex.  So I wonder, instead, if it wouldn't be good to have a `schedule()` function in this module, that took as argument the time (as a number of seconds or a number of various units) and a callback to call when the time is up.  Would it be useful?  I mean, beyond my own needs?
- I'm a bit doubtful about chain_ events.  I really feel like it's useful for builders/wizards/immortals to create events that are not called by the system at any time, but are needed for their own purpose.  And, as pointed out, it's the best way to create a pause between two portions of the event, which is useful in many cases.  Without this feature, and the `call()` function with an optional delay, it would be impossible to have pauses.  Though I looked for some funcitons like `wait(X)` that one could write in the middle of the code, but that seems difficult to handle (it cannot freeze everything, and we can't know the level of complexity of the `wait()` function, it could be in a `if` or even a `for`, handling these cases might prove too complicated).
- Regarding helpers, I did some testing yesterday, to create objects for instance.  Specifying the typeclass each time seems a bit too technical considering some paths are long and the ones using the event system don't really have to know the code by heart.  It's optional if we have a create_* helper function, and I guess we can leave this to developers willing to use such a system.  Like create_room or create_exit.  But again, that's just a suggestion.
- The example you gave of inline functions does hide the process... but it also means, if I'm not mistaken, that the text displayed at every time is statically-defined in the code.  I was looking at some feature that would allow to write dynamic description that would be different in every room.  A street is deserted at night.  A forest might be a bit scary.  A desert might be beautiful.  You see the idea.

That's all, we're about to come up with something usable, I believe, and that feels great!  I will begin to look on the EvEditor PR, which can be done right now.  I'll wait for your approval before coding the `schedule()` function in the time system.

**Comment by Griatch** [comment]
- Something like `schedule` would be useful. It doesn't exactly match the use case of either TickerHandler ticker nor a Script since both of those operate by time difference. Since the distance to a given time will be pretty much individual to each "cron job" I guess a Script may be the best bet. 
- What you are describing is not an event but a personalized callback function. Considering you are already executing arbitrary code, it would not be hard to have a user with its own "library" of string-form callbacks they could store on themselves: `self.db.my_callbacks = {"myfunc": "def blah(caller, foo): caller.db.foo = foo"}`. Then they just use some helper `delay(30, "myfunc")`, which simply fetches the callbacks and execs it... something like that. The problem with this is that it's easy to start to use this power to call other functions (say with a time of 0). Once builders start to build up their own "library" of code (and you know they will) it will very quickly veer into spaghetti code that is very hard to read, maintain and vet. 
- I would say that which typeclasses are available is highly dependent on the game. Since helpers are expandable by the developer such things should be left up for them; one can just supply a few examples. 
- The return is not statically defined in the code. It is a function that is being called, with the session supplied by Evennia and the arguments supplied by the builder/immortal (in this case). If you want the return value be influenced by a game state or by running some builder/immortal-defined code, that's up to you. The builder/immortal cannot supply new inlinefuncs from in-game, but it doesn't mean you couldn't supply ones that are themselves dynamic in various ways.

**Comment by vlegoff** [comment]
- I'm going to work on `schedule` although I'm not yet sure how to handle the case.  It seems the callback will have to be defined once and only once.  Scripts seem better from what I've seen, but I'll have to think again about the feature and do some testing.
- Functions could be set here.  There are a few drawbacks to having users creating their own functions in events:  first of all, the namespace would be restricted to the event creating it.  That is, if you want a "callback" that would be accessible through different events, that's not very easy to set, unless the callback is stored somehow.  Which is the second problem: `delay()` will not store the callback in case of a `reload()`, since it's not a top-level function, which is, in my opinion, a problem.  Actions could be interrupted quite easily and remain in a state that would require additional debug... like a NPC staying in a room between its origin and destination.  I don't think it's a good choice here.  That's why "chain_" events or whatever you want to call them could fix the issue:  they would be like callbacks, actually, user-defined, but would be shared and stored easily.
- Let's have the developers add their own helpers then, it's the most simple choice.  The user of events could, by default, use the Evennia flat API (or a limited portion of it without importing anything).  If the developer wants additional features, he/she can add them (I would personally have helpers to create static objects, rooms and exits without requiring the typeclass, I find it prettier, but that's just an opinion :P ).
- I've talked in details with @BlauFleuer about this.  I think I'll remove events in description from the PR.  I will probably use it in my own game, because they seem to allow cases that are not handled, or not handled easily, by inlinefuncs, but that's a rather rare need, and we shouldn't bother developers with it to add a layer that virtually doesn't do anything more.
And we're done to 4!

**Comment by BlauFeuer** [comment]
To points 1 and 2:

1. I think the most immediately usable and universal piece of your proposal is point 1, some sort of schedule scheme.  Currently a script or ticker would have to keep checking a timestamp to see if an event should begin, or the event would have to be triggered by some action of the player.

1. Regarding point 2, you could grow a library of call-back functions and have the builder either create a new one or use an existing one if you wanted an event to share the same callback function.   As for `delay()` not persisting callbacks, there are some techniques you can use to mitigate that, or use tickers, which restart after a`@reload` or stop/start, that remove themselves after the full event chain has transpired, or opt to use script objects which can persist through a `@reload` or stop/start.

**Comment by vlegoff** [comment]
Okay.  I'd admit I'm not sure why you dislike the idea of chain events.  That'll add a new helper function, like `deny()`, with no other purpose, but it saves users some complexity.  You could have created a script, of course, though I'm not sure how it could easily be edited through the event system.  If you write a function in an event, unless I'm greatly mistaken, no matter what, it cannot be persisted.  So we can imagine a script with some "arbitrary code" in a string, and have the event modify it... but that sounds like a lot where chain events (or secondary events, or whatever you call them) will be easier in syntax.

The other question relies in variables.  The `call` helper function could easily freeze the current variables and send them to the script being called to execute the chained event.  One can create callbacks with parameters, of course (they are just functions, after all) but in order to have these variables, it would add a bit of code in the event itself, code that would need to be repeated everytimme you would need this callback.

In short, doing this with user-defined callbacks is possible (I guess we can even find a workaround for the non-persisting status of these functions) but it seems to add a lot of code in events to do a very simple thing.  I quite agree, we shouldn't add tons of helper functions, but with `deny()` and `call()`, we're only at two... and I think both could prove pretty useful in the end.

Regarding `schedule()` I'm going to look at it in details and see how I can best handle the situation.  I've imagined scripts being created behind the scene and calling the callback when time is up (without constantly checking for a timestamp) but I have to spend some more time reading the module for game time first.

**Comment by vlegoff** [comment]
The developer documentation has been rewritten.  There's not a lot of help on how to use this system (it will be kept in a separate document) but there are more indications on how the event system works, from the perspective of developers.  I hope that's useful, feedback are welcome:  https://github.com/vlegoff/evennia/tree/events/evennia/contrib/events

**Comment by vlegoff** [comment]
The proposed user documentation can be read at: https://github.com/vlegoff/evennia/blob/events/evennia/contrib/events/USERDOC.md

It addresses some concerns raised by Griatch and others.  Some features have been removed, others are more Evennia-ish in Design.  The number of helpers has been decreased.  It seems a better, more useful system now.

As always, thanks for your feedback!

**Comment by vlegoff** [comment]
Closing this issue, as the discussion continues on a dedicated pull request.
